### PR TITLE
Add snapshotter extension 

### PR DESCRIPTION
**Pull request by trinitys7**
## Description

This PR is to add the snapshotter extension to sync wasm directory :

- Move default wasm directory from runtime dictionary into chain home folder, `ibc_08-wasm_client_data` will be subfolder of ~/.simapp
- Add IteratorCode function to iterate wasmcode that saved in chain state.
- Testing:  setup testing app with snapshot extension enable.

closes: #4275

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: testing/simapp/test_helpers.go [hunk 1]
```diff
@@ -113,6 +114,98 @@ func SetupWithGenesisValSet(t *testing.T, valSet *cmttypes.ValidatorSet, genAccs
 	return app
 }
 
+func setup(tb testing.TB, chainID string, withGenesis bool, invCheckPeriod uint) (*SimApp, GenesisState) {
+	tb.Helper()
+	db := dbm.NewMemDB()
```
Comment (trinitys7) [inline testing/simapp/test_helpers.go:119]: Those testing functions are from the current `initSetup`, I added some app options to run statesync with the testing app.

**Comment by faddat** [comment]
thank you

**Comment by damiannolan** [comment]
Thank you @trinitys7 for this PR. The code looks great so far. I'm thinking that this will probably need to wait temporarily until https://github.com/cosmos/ibc-go/pull/4292 is merged first. 

Is it possible for you to click the `Allow edits from maintainers` button which will allow us to keep the branch in sync and push any updates needed.

**Comment by trinitys7** [comment]
@damiannolan because this branch from my organization, it's unable to `Allow edits from maintainers`. I'll close this PR and make new PR from my personal account for it.

**Pull request closed without merging**